spec-0007/index.md

```diff
@@ -0,0 +1,183 @@
+    if rng is not _NoValue:
+        return np.random.default_rng(rng)
+
+    if seed is _NoValue:
+        # If the user passed nothing, we have to reach into NumPy here:
+        # 1. If np.random.seed(None) was called (or never called), then we can
+        #    just use the default_rng (the result is random anyway).
+        # 2. If it was called, we must return the global random state object
+        #    and warn about future ignoring of seed!
+        if np.random.mtrand._rand._bit_generator._seed_seq is not None:
+            # The user did not seed, so no need to warn.
+            return np.random.default_rng()
+        warnings.warn(
+            "The NumPy global rng was seeded in call to np.random.seed() "
+            "in the future this function will ignore this seed and return "
+            "random values as if a new `np.random.default_rng()` was created.",
+            FutureWarning, stacklevel=5)
+        return np.random.mtrand._rand
+    if seed is None or seed is np.random:
+        return np.random.mtrand._rand
+    if isinstance(seed, (numbers.Integral, np.integer)):
+        return np.random.RandomState(seed)
+    if isinstance(seed, (np.random.RandomState, np.random.Generator)):
+        return seed
+
+    raise ValueError(f"'{seed}' cannot be used to seed a numpy.random.RandomState"
+                     " instance")
+
+
+def _prepare_rng(old_name, dep_version=None):
+    new_name = "rng"
+
+    def decorator(fun):
+        @functools.wraps(fun)
+        def wrapper(*args, **kwargs):
+            if old_name in kwargs:
+                if dep_version:
+                    end_version = dep_version.split('.')
+                    end_version[1] = str(int(end_version[1]) + 2)
+                    end_version = '.'.join(end_version)
+                    message = (f"Use of keyword argument `{old_name}` is "
+                               f"deprecated and replaced by `{new_name}`.  "
+                               f"Support for `{old_name}` will be removed "
+                               f"in SciPy {end_version}.")
+                    warnings.warn(message, DeprecationWarning, stacklevel=2)
+                if new_name in kwargs:
+                    message = (f"{fun.__name__}() got multiple values for "
+                               f"argument now known as `{new_name}`")
+                    raise TypeError(message)
+
+            kwargs[new_name] = check_random_state(
+                kwargs.pop(old_name, _NoValue),
+                rng=kwargs.pop(new_name, _NoValue)
+            )
+            return fun(*args, **kwargs)
+        return wrapper
+    return decorator
+
+
+@_prepare_rng("random_state")
+def library_function(/, rng=None):
+    # The decorated library function takes an `rng` argument which is
+    # guaranteed to be a either a Generator or a RandomState.
+    # `random_state=` is supported input (the old can be customized).
+    assert isinstance(rng, (np.random.Generator, np.random.RandomState))
```

Comment:
Just some optional feedback for this example since I recently wrote a blog post on precisely that: [Deprecate positional arguments with placeholders](https://grueter.dev/blog/deprecate-arg-with-placeholder/). That approach has the advantage that the deprecated parameter still shows up in the signature during the deprecation. If wanted I'm happy to come up with an updated version of this snippet. :wink:

Reply:
Oh, I don't mean to hold this up. We can always update after the merge.

Reply:
That was what we actually wanted 😅 To not show the deprecated usage. There are various opinions there OC. My view as a user is that I don't like to have to go through all the API doc and pay attention to see if one field is deprecated or not, i.e. just give me the recommended way of doing things.

This is either way just a snippet that project can use as a Quickstart and adapt to their own policy.

Reply:
I agree with Pamphile here and try to remove the deprecated parameters from the docstrings as soon as the deprecation happens to help the transition, and help to avoid new usage of deprecated ways.

Reply:
Any code updates you'd like to see here, then?

Reply:
This code doesn't seem to do anything about the docstrings, so should be fine?

Though a super explicit sentence about signatures and docstrings may be useful, as the code example really goes down to the weeds of how one does deprecations (e.g. there are tons of [deprecation decorators out there](https://docs.astropy.org/en/latest/api/astropy.utils.decorators.deprecated_renamed_argument.html#astropy.utils.decorators.deprecated_renamed_argument), I don't necessarily see why this code snipped is needed in the SPEC). 
Anyway, I still miss something along the lines of:

-  `random_seed` should be removed from signatures and docstrings in favour of `rng`.

Reply:
I think that's a step beyond the scope of what a SPEC should attempt do.

Reply:
@rkern Do you mean you think it's best to not give a code example?

Reply:
btw, my comments should not block this to be merged as draft as is.

Reply:
I mean that it's best not to recommend what should/should not be documented.

Reply:
Having read the above, I'd say it's fine to include a snippet. However, it should make it very clear in which situations this can be used.

To add a bit of context to my suggestion. I think this example assumes that the deprecated parameter is keyword-only. For libraries that expose the deprecated `random_state`/`seed` as a positional parameter, this snippet wouldn't help / handle that correctly.

Hence my suggestion above to use a placeholder. Without, the signature doesn't reflect the behavior. Very confusing if a user tries to target a positional parameter after the intended input is swallowed by the deprecated one. That limitation should be addressed either as a comment or by patching the snippet?

Reply:
I would just be clear at the beginning that this snippet is, well a snippet. i.e. just an aid to help a maintainer quick-start things on their repo.

Reply:
@mdhaber Is reviewing the text surrounding the snippet now.